smithy-model/src/main/java/software/amazon/smithy/model/transform/CreateDedicatedInputAndOutput.java

```diff
@@ -0,0 +1,220 @@
+        } else {
+            return createSyntheticShape(model, operation, inputSuffix, input, new InputTrait());
+        }
+    }
+
+    private StructureShape createdUpdatedOutput(
+            Model model,
+            OperationShape operation,
+            StructureShape output,
+            NeighborProvider reverse
+    ) {
+        if (output.hasTrait(OutputTrait.class)) {
+            return renameShapeIfNeeded(model, output, operation, outputSuffix);
+        } else if (isDedicatedHeuristic(operation, output, reverse, RelationshipType.OUTPUT)) {
+            OutputTrait trait = new OutputTrait();
+            return renameShapeIfNeeded(model, output.toBuilder().addTrait(trait).build(), operation, outputSuffix);
+        } else {
+            return createSyntheticShape(model, operation, outputSuffix, output, new OutputTrait());
+        }
+    }
+
+    private StructureShape renameShapeIfNeeded(
+            Model model,
+            StructureShape struct,
+            OperationShape operation,
+            String suffix
+    ) {
+        // Check if the shape already has the desired name.
+        ShapeId expectedName = ShapeId.fromParts(operation.getId().getNamespace(),
+                                                 operation.getId().getName() + suffix);
+        if (struct.getId().equals(expectedName)) {
+            return struct;
+        }
+
+        ShapeId newId = createSyntheticShapeId(model, operation, suffix);
+        return struct.toBuilder()
+                .id(newId)
+                .addTrait(new OriginalShapeIdTrait(struct.getId()))
+                .build();
+    }
+
+    private boolean isDedicatedHeuristic(
+            OperationShape operation,
+            StructureShape struct,
+            NeighborProvider reverse,
+            RelationshipType expected
+    ) {
+        // Only assume that a shape is dedicated to the operation its name starts with the operation name.
+        if (!struct.getId().getName().startsWith(operation.getId().getName())) {
+            return false;
+        }
+
+        // Check if the shape is only referenced as input or output.
+        return isSingularReference(reverse, struct, expected);
+    }
+
+    private boolean isSingularReference(NeighborProvider reverse, Shape shape, RelationshipType expected) {
+        int totalDirectedEdges = 0;
+
+        // We need to exclude inverted edges like MEMBER_CONTAINER, and only look for directed
+        // edges like INPUT and OUTPUT.
+        for (Relationship rel : reverse.getNeighbors(shape)) {
+            if (rel.getRelationshipType().getDirection() == RelationshipDirection.DIRECTED) {
+                totalDirectedEdges++;
+                if (rel.getRelationshipType() != expected) {
+                    return false;
+                }
+            }
+        }
+
+        return totalDirectedEdges == 1;
+    }
+
+    private static StructureShape createSyntheticShape(
```

Comment:
I don't think this would work with the `Unit` type since the unitType trait can only be applied to it.

Reply:
Great catch. Fixed.